korma/src/commonMain/kotlin/com/soywiz/korma/interpolation/Easing.kt

```diff
@@ -2,92 +2,171 @@ package com.soywiz.korma.interpolation
 
 import kotlin.math.*
 
+private inline fun combine(it: Double, start: Easing, end: Easing) =
+    if (it < 0.5) 0.5 * start(it * 2.0) else 0.5 * end((it - 0.5) * 2.0) + 0.5
+private const val BOUNCE_FACTOR = 1.70158
+private const val HALF_PI = PI / 2.0
+
 @Suppress("unused")
 fun interface Easing {
     operator fun invoke(it: Double): Double
 
     companion object {
         fun cubic(f: (t: Double, b: Double, c: Double, d: Double) -> Double): Easing = Easing { f(it, 0.0, 1.0, 1.0) }
-        fun combine(start: Easing, end: Easing) = Easing { if (it < 0.5) 0.5 * start(it * 2.0) else 0.5 * end((it - 0.5) * 2.0) + 0.5 }
+        fun combine(start: Easing, end: Easing) = Easing { combine(it, start, end) }
 
-        operator fun invoke(f: (Double) -> Double) = Easing { it -> f(it) }
+        operator fun invoke(f: (Double) -> Double) = Easing { f(it) }
 
-        val SMOOTH get() = Easings.SMOOTH
-        val EASE_IN_ELASTIC get() = Easings.EASE_IN_ELASTIC
-        val EASE_OUT_ELASTIC get() = Easings.EASE_OUT_ELASTIC
-        val EASE_OUT_BOUNCE get() = Easings.EASE_OUT_BOUNCE
-        val LINEAR get() = Easings.LINEAR
-        val EASE_IN get() = Easings.EASE_IN
-        val EASE_OUT get() = Easings.EASE_OUT
-        val EASE_IN_OUT get() = Easings.EASE_IN_OUT
-        val EASE_OUT_IN get() = Easings.EASE_OUT_IN
-        val EASE_IN_BACK get() = Easings.EASE_IN_BACK
-        val EASE_OUT_BACK get() = Easings.EASE_OUT_BACK
-        val EASE_IN_OUT_BACK get() = Easings.EASE_IN_OUT_BACK
-        val EASE_OUT_IN_BACK get() = Easings.EASE_OUT_IN_BACK
-        val EASE_IN_OUT_ELASTIC get() = Easings.EASE_IN_OUT_ELASTIC
-        val EASE_OUT_IN_ELASTIC get() = Easings.EASE_OUT_IN_ELASTIC
-        val EASE_IN_BOUNCE get() = Easings.EASE_IN_BOUNCE
-        val EASE_IN_OUT_BOUNCE get() = Easings.EASE_IN_OUT_BOUNCE
-        val EASE_OUT_IN_BOUNCE get() = Easings.EASE_OUT_IN_BOUNCE
-        val EASE_IN_QUAD get() = Easings.EASE_IN_QUAD
-        val EASE_OUT_QUAD get() = Easings.EASE_OUT_QUAD
-        val EASE_IN_OUT_QUAD get() = Easings.EASE_IN_OUT_QUAD
-        val EASE_SINE get() = Easings.EASE_SINE
+        @Deprecated("Moved to Easings enum", ReplaceWith("Easings.SMOOTH"))
+        val SMOOTH: Easing = Easings.SMOOTH
```

Comment:
Misses `get() =` that might causes issues on initialization order.

In fact, initially I put the implementation here directly without the private class Easings, but had to split it and use getters and failed, not sure if on JS or on Kotlin/Native due to initialization order. For example in JS I getting `unefined` on some cases at runtime.

Reply:
Actually, I did run into one issue that might be related to this. Initially I didn't have the extra `combine` method and instead stored a `Easing.combine(EASE_IN, EASE_OUT)`-type thing in the Easings that were derived from other Easings. This, confusingly, led to NPE's in the enum constructor, which _might_ be relaetd to what you're talking about.

> might 

Would existing tests catch these?

Reply:
I'm not 100% sure, but I think initialization order is something that might cause issues not always but in some cases. When I wrote JTransc and tried to do static initialization found some issues regarding to cycles in the objects that depending on the tree shaking / dead code elimination (classes and methods included) caused issues sometimes and was avoided others by chance.

In the case of Easing I remember having those problems but can't tell you the details because I simply don't remember them